### PR TITLE
Largely functional DB query interface

### PR DESCRIPTION
**Pull request by rtilder**
memcache updates need to be flushed out and invalid token responses
need to be sorted out; they're breaking the functional tests for some
odd reason I haven't been able to fathom yet.

### CODE REVIEW

File: syncstorage/storage/__init__.py [hunk 1]
```diff
@@ -307,6 +313,63 @@ def delete_items(self, userid, collection, items):
...
+            The new transaction's ID as a BigInteger timestamp.
+        """
+
+    @abc.abstractmethod
+    def append_items_to_transaction(self, batchid, userid, collection, items):
+        """Creates or updates multiple items from a multi-POST batch to a
+        transaction.
+
+        Args:
+            batchid: big integer batch identifier for this transaction
```
Comment (rfk) [inline syncstorage/storage/__init__.py]: The naming confusion between "batchid" and "transaction" really stands out in the code here; if the name of the thing is to be "transaction" then lets make this "trnid" or "transactionid" or something all throughout the code and the API.

File: syncstorage/storage/__init__.py [hunk 2]
```diff
@@ -307,6 +313,63 @@ def delete_items(self, userid, collection, items):
...
+            userid: integer identifying the user in the storage.
+            collection: name of the collection.
+            items: a list of dicts giving data for each item.
+
+        Returns:
+            The new last-modified timestamp for the transaction.
```
Comment (rfk) [inline syncstorage/storage/__init__.py]: It's not obvious to me why this timestamp would be useful, and if we're not going to use it then perhaps we should just return `None`.

File: syncstorage/storage/__init__.py [hunk 3]
```diff
@@ -307,6 +313,63 @@ def delete_items(self, userid, collection, items):
...
+
+        Raises:
+            ConflictError: the operation conflicted with a concurrent write.
+        """
+
+    @abc.abstractmethod
+    def commit_transaction(self, batchid, userid, collection):
```
Comment (rfk) [inline syncstorage/storage/__init__.py]: For consistency with the other methods on this class, perhaps the order of arguments here should be `userid`, `collection`, `batchid`.

File: syncstorage/views/__init__.py [hunk 4]
```diff
@@ -295,6 +315,9 @@ def post_collection(request):
     bsos = request.validated["bsos"]
     invalid_bsos = request.validated["invalid_bsos"]
 
+    if request.validated["batch"] or request.validated["commit"]:
+        return post_collection_transaction(request)
+
```
Comment (rfk) [inline syncstorage/views/__init__.py]: It may be worth pulling the original non-transactional implementation out into a separate `post_collection_single` function or similar, to make it more obvious that we're choosing between two independent implementations.

File: syncstorage/views/__init__.py [hunk 5]
```diff
@@ -319,6 +342,84 @@ def post_collection(request):
...
+    batch = request.validated["batch"]
+    commit = request.validated["commit"]
+
+    # The "batch" key is set only on a multi-POST batch request prior to a
+    # commit.  The "modified" key is only set upon a successful commit.
+    # The two flags are mutually exclusive.
+    # Any failures at all mean cancelling the transaction completely.
+    res = {'success': [], 'failed': {}}
+
+    # If there are any parsing failures, we won't even start a transaction.
```
Comment (rfk) [inline syncstorage/views/__init__.py]: We should check this behaviour with @rnewman, since what you describe differs from behaviour in the non-transactional case.  In that case the valid bsos are written and the invalid ones are rejected, and I can see an argument that we should do the same here, report the failures to the client, and let it decide what to do about the problem.

File: syncstorage/views/__init__.py [hunk 6]
```diff
@@ -319,6 +342,84 @@ def post_collection(request):
...
+    try:
+        if batch is True:
+            try:
+                batch = storage.create_transaction(userid, collection)
+            except ConflictError, e:
+                # ConflictError here means a client is spamming requests,
+                # I think.
+                logger.error('Collision in transaction creation!')
+                logger.error(e)
+                raise HTTPServiceUnavailable('Too many transactions opened')
```
Comment (rfk) [inline syncstorage/views/__init__.py]: I think we should raise the original ConflictError here, and it'll be converted into an appropriate response by the middleware layer (I believe it's a 503)

File: syncstorage/views/__init__.py [hunk 7]
```diff
@@ -319,6 +342,84 @@ def post_collection(request):
...
+                         '("{0}")'.format(batch))
+            logger.error(str(e))
+            for bso in bsos:
+                res["failed"][bso["id"]] = "db error"
+            raise
+
+        if commit:
+            try:
+                storage.commit_transaction(batch, userid, collection)
+                res['modified'] = ts
```
Comment (rfk) [inline syncstorage/views/__init__.py]: IIRC, we want all items in the transaction to get the timestamp from the time of commit.  It may be slightly more reliable to use `ts = store.commit_transaction()` here rather than using the timestamp from the previous `append_items_to_transaction`.

File: syncstorage/views/__init__.py [hunk 8]
```diff
@@ -319,6 +342,84 @@ def post_collection(request):
...
+                request.response.headers["X-Last-Modified"] = str(ts)
+            except ConflictError:
+                raise
+            except Exception, e:
+                logger.error("Could not commit transaction")
+                logger.error(e)
+                raise
+            finally:
+                for bso in bsos:
+                    res["failed"][bso["id"]] = "db error: commit"
```
Comment (rfk) [inline syncstorage/views/__init__.py]: Will this `finally` block trigger even if there's no exception, reporting all bsos as failed?

File: syncstorage/views/__init__.py [hunk 9]
```diff
@@ -319,6 +342,84 @@ def post_collection(request):
...
+        else:
+            res["batch"] = batch
+        res["success"].extend([bso["id"] for bso in bsos])
+    except ConflictError:
+        raise
+    finally:
+        if batch and batch is not True:
+            storage.close_transaction(batch, userid, collection)
```
Comment (rfk) [inline syncstorage/views/__init__.py]: Likewise, won't this result in `close_transaction` being called on every request except the initial one?

File: syncstorage/views/validators.py [hunk 10]
```diff
@@ -146,6 +146,31 @@ def extract_query_params(request):
...
+    is to be committed and deleted.
+    """
+    request.validated["batch"] = False
+    batch_id = request.GET.get("batch")
+    if batch_id is not None:
+        if batch_id == "true":
+            batch_id = True
+        request.validated["batch"] = batch_id
+    elif batch_id is None and "batch" in request.GET:
+        request.validated["batch"] = True
```
Comment (rfk) [inline syncstorage/views/validators.py]: Is it actually possible for `"batch"` to wind up in `request.GET` with a value of `None`?

File: syncstorage/views/validators.py [hunk 11]
```diff
@@ -146,6 +146,31 @@ def extract_query_params(request):
...
+
+    request.validated["commit"] = False
+    commit = request.GET.get("commit")
+    if commit is not None and commit == "true":
+        request.validated["commit"] = True
```
Comment (rfk) [inline syncstorage/views/validators.py]: If there's a value other than `"true"` we should probably return a 400 error.

File: syncstorage/views/__init__.py (hunk 5 above)
Comment (rnewman) [inline syncstorage/views/__init__.py]: This is an aspect of the Sync protocol that I kinda hate: Sync essentially operates like `INSERT OR IGNORE`, including for malformed input, which is so rarely what one wants!

Fortunately, in the case of batched uploads, the client can see that some of the uploaded BSOs were rejected, and can choose to not commit the entire transaction. 

I think it's fine to keep the existing behavior (queue the ones that were fine) for that reason; if anything, the sane thing to do the reverse: to reject part-wrong non-batch `POST`s and part-accept batched ones!

Two observations:

- It'd be nice to have a `?discard` operation, just for efficiency's sake.
- Careful clients should always issue an empty `POST` to commit, so that they don't accidentally commit a partial final batch.

File: syncstorage/tests/functional/test_storage.py [hunk 12]
```diff
@@ -1286,6 +1287,49 @@ def test_pagination_with_newer_and_sort_by_oldest(self):
...
+        self.app.post_json(endpoint, bsos)
+
+        bso3 = {'id': 'a', 'payload': 'internal'}
+        bso4 = {'id': 'b', 'payload': 'pancreas'}
+        resp = self.app.post_json(endpoint + '?batch=true', [bso3, bso4])
+        batch = resp.json["batch"]
+        # This is sooooo going to break when it's run at just the wrong time
+        # and will result in a seriously confused person at the keyboard.
+        guess = float(resp.headers["X-Weave-Timestamp"])
+        self.assertEqual(batch / 1000, math.floor(guess))
```
Comment (rfk) [inline syncstorage/tests/functional/test_storage.py]: Since the value of batch is defined to be opaque, we probably don't need to check that it has any particular value, but rather just check that it's an appropriate-looking string.

File: syncstorage/tests/functional/test_storage.py [hunk 13]
```diff
@@ -1286,6 +1287,49 @@ def test_pagination_with_newer_and_sort_by_oldest(self):
...
+
+        # make sure we don't have the pending commit
+        resp = self.app.get(endpoint)
+        res = resp.json
+        res.sort()
+        self.assertEquals(res, ['12', '13'])
+        self.assertEquals(int(resp.headers['X-Weave-Records']), 2)
```
Comment (rfk) [inline syncstorage/tests/functional/test_storage.py]: It'd also be worth checking that the timestamp on the collection hasn't been modified yet.

File: syncstorage/tests/functional/test_storage.py [hunk 14]
```diff
@@ -1286,6 +1287,49 @@ def test_pagination_with_newer_and_sort_by_oldest(self):
...
+        res = resp.json
+        res.sort()
+        self.assertEquals(res, ['12', '13', 'a', 'b', 'c'])
+        self.assertEquals(int(resp.headers['X-Weave-Records']), 5)
+        resp = self.app.get(endpoint + '/13')
+        self.assertEquals(resp.json["payload"], 'portnoy')
+
+        # invalid transaction ID
+        bso7 = {'id': 'd', 'payload': 'burrito'}
+        resp = self.app.post_json(endpoint + '?batch=sammich', [bso7])
```
Comment (rfk) [inline syncstorage/tests/functional/test_storage.py]: Please break these error-case tests out into one or more separate test methods.

File: syncstorage/tests/functional/test_storage.py [hunk 15]
```diff
@@ -1286,6 +1287,49 @@ def test_pagination_with_newer_and_sort_by_oldest(self):
...
+                          float(resp.headers['X-Weave-Timestamp']))
+
+        # make sure the changes apply
+        resp = self.app.get(endpoint)
+        res = resp.json
+        res.sort()
+        self.assertEquals(res, ['12', '13', 'a', 'b', 'c'])
+        self.assertEquals(int(resp.headers['X-Weave-Records']), 5)
+        resp = self.app.get(endpoint + '/13')
+        self.assertEquals(resp.json["payload"], 'portnoy')
```
Comment (rfk) [inline syncstorage/tests/functional/test_storage.py]: There are probably additional consistency checks worth applying here, e.g. checking that all the items in the batch got assigned the same timestamp, and that it matches the timestamp reported in the `commit` response above.

Comment (rfk) [inline syncstorage/tests/functional/test_storage.py]: In fact, @rnewman, I encourage you to think of any specific edge-case-style interactions that might affect this API, and we can add specific tests for them.

File: syncstorage/storage/sql/__init__.py [hunk 16]
```diff
@@ -456,6 +456,54 @@ def set_items(self, session, userid, collection, items):
...
+    def append_items_to_transaction(self, session, batchid, userid,
+                                    collection, items):
+        """Inserts items into batch_upload_items"""
+        rows = []
+        for data in items:
+            id_ = data["id"]
+            row = self._prepare_bui_row(session, batchid, id_, data)
+            rows.append(row)
+        defaults = {"modified": ts2bigint(session.timestamp)}
+        session.insert_or_update("batch_upload_items", rows, defaults)
```
Comment (rfk) [inline syncstorage/storage/sql/__init__.py]: Per https://bugzilla.mozilla.org/show_bug.cgi?id=1250189#c16, I suspect we want all items in the transaction to receive the same modification time, and for that time to be the time of the final commit.

File: syncstorage/storage/sql/dbconnect.py [hunk 17]
```diff
@@ -112,19 +112,33 @@ def __new__(cls, *args, **kwds):
...
-               autoincrement=False),
-        Column("collection", Integer, primary_key=True, nullable=False,
-               autoincrement=False),
-        Column("id", String(64), primary_key=True, autoincrement=False),
         Column("sortindex", Integer),
         Column("modified", BigInteger, nullable=False),
+        # MySQL no likey "DEFAULT ''".  Results in an error:
+        #     ERROR 1101 (42000): BLOB, TEXT, GEOMETRY or JSON column 'payload'
+        #                         can't have a default value
+        # Column("payload", PAYLOAD_TYPE, nullable=False),
```
Comment (rfk) [inline syncstorage/storage/sql/dbconnect.py]: Ouch, not sure why we haven't hit this before :-/.  I guess we may as well fix it as part of this work, unless it turns out to be a rabbit-hole.

File: syncstorage/storage/sql/dbconnect.py [hunk 18]
```diff
@@ -144,21 +159,85 @@ def _get_bso_columns(table_name):
...
+        Column("item", String(64), primary_key=True, nullable=False,
+               autoincrement=False)
+    )
+    common = _common_columns(MAX_TTL)
+    indexes = (
+        # Declare indexes.
+        # We need to include the tablename in the index name due to sharding,
+        # because index names in sqlite are global, not per-table.
+        # Index on "ttl" for easy pruning of expired items.
+        Index("%s_ttl_idx" % (table_name,), "ttl"),
```
Comment (rfk) [inline syncstorage/storage/sql/dbconnect.py]: I'm not convinced we'll ever want to prune expired items from this table, but rather just prune everything that's been committed.  In fact, do we need anything other than the primary key index on this table?

File: syncstorage/storage/sql/dbconnect.py [hunk 19]
```diff
@@ -144,21 +159,85 @@ def _get_bso_columns(table_name):
 
 bso = Table("bso", metadata, *_get_bso_columns("bso"))
 
+# Table mapping (user_id, collection_id) => batch IDs
+
+batch_uploads = Table(
+    "batch_uploads",
+    metadata,
+    Column("batch", BigInteger, primary_key=True, nullable=False),
```
Comment (rfk) [inline syncstorage/storage/sql/dbconnect.py]: If we're using just a timestamp as the batch id, we'll almost certainly generate spurious collisions between multiple users trying to create a transaction at the same time.  We could make this a compound primary key on `(batch, userid)`, or make `batch` a string and include more information that way.  @rtilder thoughts?

File: syncstorage/storage/sql/dbconnect.py [hunk 20]
```diff
@@ -144,21 +159,85 @@ def _get_bso_columns(table_name):
...
+    Column("userid", Integer, nullable=False, autoincrement=False),
+    Column("collection", Integer, nullable=False, autoincrement=False),
```
Comment (rfk) [inline syncstorage/storage/sql/dbconnect.py]: You shouldn't need `autoincrement=False` on these non-primary-key columns.

File: syncstorage/storage/sql/dbconnect.py [hunk 21]
```diff
@@ -371,6 +454,11 @@ def get_query(self, name, params):
                 qvars["bso"] = params["bso"]
             else:
                 qvars["bso"] = self.get_bso_table(params["userid"])
+        if "%(batch_upload_items)s" in query:
```
Comment (rfk) [inline syncstorage/storage/sql/dbconnect.py]: nit: the inconsistency between abbreviated `%(bso)s` and long-form `%(batch_upload_items)s` is a little strange.

File: syncstorage/storage/sql/dbconnect.py [hunk 22]
```diff
@@ -387,6 +475,12 @@ def get_bso_table(self, userid):
             return bso
         return get_bso_table(userid % self.shardsize)
 
+    def get_batch_item_table(self, userid):
+        """Get the batch_upload_items table object for the given userid."""
+        if not self.shard or userid is None:
+            return bui
+        return get_bso_table(userid % self.shardsize)
```
Comment (rfk) [inline syncstorage/storage/sql/dbconnect.py]: s/get_bso_table/get_batch_item_table/g

File: syncstorage/storage/sql/queries_generic.py [hunk 23]
```diff
@@ -12,6 +12,7 @@
 string interpolation variables with special meaning:
 
     * %(bso)s:   insert the name of the user's sharded BSO storage table
+    * %(bui)s:   insert the name of the user's sharded batch_upload_items table
```
Comment (rfk) [inline syncstorage/storage/sql/queries_generic.py]: documentation doesn't match code here, code uses `%(batch_upload_items)s`

File: syncstorage/storage/sql/queries_generic.py [hunk 24]
```diff
@@ -92,6 +93,47 @@
...
+                     "   payload, payload_size, ttl) "\
+                     "SELECT "\
+                     "  :userid, :collection, item, sortindex, modified, "\
+                     "  COALESCE(payload, \"\"), COALESCE(payload_size, 0), "\
+                     "  COALESCE(ttl, ttl + :default_ttl) "\
+                     "FROM %(batch_upload_items)s "\
+                     "WHERE batch = :batch "\
+                     "ON DUPLICATE KEY UPDATE "\
+                     "  sortindex = COALESCE(VALUES(%(bso)s.sortindex), " \
+                     "                   %(batch_upload_items)s.sortindex), " \
```
Comment (rfk) [inline syncstorage/storage/sql/queries_generic.py]: The order of arguments to `COALESCE` seems off here...IIUC we want to take the value from `batch_upload_items` if it's non-NULL, but fall back to keeping the value from `bso` if it is NULL.  So the arguments should be the other way around.

File: syncstorage/storage/sql/queries_generic.py [hunk 25]
```diff
@@ -92,6 +93,47 @@
...
+COMMIT_TRANSACTION = "INSERT INTO %(bso)s "\
+                     "  (userid, collection, id, sortindex, modified, "\
+                     "   payload, payload_size, ttl) "\
+                     "SELECT "\
+                     "  :userid, :collection, item, sortindex, modified, "\
+                     "  COALESCE(payload, \"\"), COALESCE(payload_size, 0), "\
+                     "  COALESCE(ttl, ttl + :default_ttl) "\
+                     "FROM %(batch_upload_items)s "\
+                     "WHERE batch = :batch "\
+                     "ON DUPLICATE KEY UPDATE "\
```
Comment (rfk) [inline syncstorage/storage/sql/queries_generic.py]: Since `ON DUPLICATE KEY UPDATE` is a MySQL-specific extension, this version of the query should go into `queries_mysql` and a more generic one should be used here.  Is the sqlite version generic enough to work as the default implementation?

File: syncstorage/storage/sql/queries_generic.py [hunk 26]
```diff
@@ -163,3 +205,8 @@ def FIND_ITEMS(bso, params):
 # version using DELETE <blah> LIMIT 1000.
 PURGE_SOME_EXPIRED_ITEMS = "DELETE FROM %(bso)s "\
                            "WHERE ttl < (UNIX_TIMESTAMP() - :grace) "
+
+PURGE_TRANSACTIONS = "DELETE FROM batch_uploads WHERE batch < :batch"
+
+PURGE_TRANSACTION_CONTENTS = "DELETE FROM batch_upload_items WHERE batch < " \
+                             ":batch"
```
Comment (rfk) [inline syncstorage/storage/sql/queries_generic.py]: Looks good, we'll also need a script to run periodically that will perform this purge.

File: syncstorage/storage/sql/queries_generic.py [hunk 27]
```diff
@@ -92,6 +93,47 @@
 DELETE_ITEMS = "DELETE FROM %(bso)s WHERE userid=:userid "\
                "AND collection=:collectionid AND id IN %(ids)s"
 
+CREATE_TRANSACTION = "INSERT INTO batch_uploads (batch, userid, collection) "\
+                     "VALUES (:batch, :userid, :collection)"
+
+VALID_TRANSACTION = "SELECT open FROM batch_uploads WHERE batch = :batch " \
+                    "AND userid = :userid AND collection = collection"
```
Comment (rfk) [inline syncstorage/storage/sql/queries_generic.py]: This isn't used anywhere yet.

In general, what's the advantage of setting `open=0` as opposed to e.g. just deleting the entry from `batch_uploads` once we're done with it?

File: syncstorage/storage/sql/__init__.py [hunk 28]
```diff
@@ -456,6 +456,54 @@ def set_items(self, session, userid, collection, items):
...
+    @with_session
+    def commit_transaction(self, session, batchid, userid, collection):
+        collectionid = self._get_collection_id(session, collection)
+        params = {
+            "batch": batchid,
+            "userid": userid,
+            "collection": collectionid,
+            "default_ttl": MAX_TTL
+        }
+        session.query("COMMIT_TRANSACTION", params)
```
Comment (rfk) [inline syncstorage/storage/sql/__init__.py]: I wonder whether it's worth adding an explicit timer around this new and scary query, so we can monitor it more closely than the mysql slow-query log might allow.

File: syncstorage/storage/sql/dbconnect.py (hunk 19 above)
Comment (rtilder) [inline syncstorage/storage/sql/dbconnect.py]: D'oh!  It's supposed to be PRIMARY KEY (batch, userid).  Looks like this plus the autoincrement=False on the collection column are victims are failed copying and pasting.

File: syncstorage/storage/sql/dbconnect.py (hunk 18 above)
Comment (rtilder) [inline syncstorage/storage/sql/dbconnect.py]: Nope.  It's just deliberate mimicry of the bso table indexes.  No real need.

File: syncstorage/storage/sql/dbconnect.py (hunk 17 above)
Comment (rtilder) [inline syncstorage/storage/sql/dbconnect.py]: There's my.ini setting to honor MySQL 4.x semantics that I suspect is more common than not.  I'll find the specifics in my notes and see if I can figure out what TravisCI's default setting is.

File: syncstorage/views/__init__.py (hunk 5 above)
Comment (rtilder) [inline syncstorage/views/__init__.py]: Discard entire group[1] of batches?

1. Should be murder of batches.

File: syncstorage/storage/sql/dbconnect.py (hunk 19 above)
Comment (rtilder) [inline syncstorage/storage/sql/dbconnect.py]: Additional thoughts: changing the type for `batch` means we can use version 4 UUIDs as identifiers directly without dealing with converting to and from integers to store them.  I'm not certain if that would affect the `batch_upload_items` table's purging performance.  I believe it wouldn't but would definitely want to verify experimentally first.

File: syncstorage/storage/__init__.py [hunk 29]
```diff
@@ -47,6 +47,12 @@ class InvalidOffsetError(StorageError, ValueError):
     pass
 
 
+class InvalidBatch(StorageError, ValueError):
```
Comment (rfk) [inline syncstorage/storage/__init__.py]: naming nit: this should be with `InvalidBatchError` for consistency with its sibling class names.

File: syncstorage/storage/memcached.py [hunk 30]
```diff
@@ -880,6 +931,85 @@ def del_item(self, userid, item):
...
+        batchid = int(batch * 1000)
+        if not bdata:
+            bdata = {}
+        if batchid in bdata:
+            raise ConflictError
+        bdata[batchid] = {"user": userid,
+                          "modified": batch,
+                          # FIXME Rough guesstimate of the maximum
+                          #       reasonable life span of a batch
+                          "expires": int(batch) + 2 * 3600,
```
Comment (rfk) [inline syncstorage/storage/memcached.py]: We should read this from config, and use it uniformly in both backends

File: syncstorage/storage/sql/dbconnect.py [hunk 31]
```diff
@@ -112,19 +112,39 @@ def __new__(cls, *args, **kwds):
...
+        # into a warning:
+        #
+        # [mysqld]
+        # sql_mode=MYSQL40
+        #
         Column("payload", PAYLOAD_TYPE, nullable=False, server_default=""),
         Column("payload_size", Integer, nullable=False,
                server_default=sqltext("0")),
-        Column("ttl", Integer, server_default=sqltext(str(MAX_TTL))),
+        Column("ttl", Integer, server_default=sqltext(str(ttl)))
```
Comment (rfk) [inline syncstorage/storage/sql/dbconnect.py]: All calls to this function pass in `MAX_TTL` as the arguments, we can probably simplify by just continuing to refer to it directly here.

File: syncstorage/storage/sql/queries_generic.py [hunk 32]
```diff
@@ -92,6 +93,46 @@
 DELETE_ITEMS = "DELETE FROM %(bso)s WHERE userid=:userid "\
                "AND collection=:collectionid AND id IN %(ids)s"
 
+CREATE_BATCH = "INSERT INTO batch_uploads (batch, userid, collection) "\
+                     "VALUES (:batch, :userid, :collection)"
+
+VALID_BATCH = "SELECT batch FROM batch_uploads WHERE batch = :batch " \
+                    "AND userid = :userid AND collection = :collection"
+
+APPEND_ITEMS_TO_BATCH = "INSERT INTO %(bui)s "\
```
Comment (rfk) [inline syncstorage/storage/sql/queries_generic.py]: This query appears to be unused, I assume because of use of the `insert_or_update` helper.

File: syncstorage/storage/sql/queries_generic.py [hunk 33]
```diff
@@ -92,6 +93,46 @@
...
+                              "  payload_size = VALUES(payload_size) "\
+                              "  ttl = VALUES(ttl)"
+
+APPLY_BATCH = "INSERT INTO %(bso)s "\
+                     "  (userid, collection, id, sortindex, modified, "\
+                     "   payload, payload_size, ttl) "\
+                     "SELECT "\
+                     "  :userid, :collection, id, sortindex, :modified, " \
+                     "  COALESCE(payload, \"\"), COALESCE(payload_size, 0), "\
+                     "  COALESCE(ttl, ttl + :default_ttl) "\
```
Comment (rfk) [inline syncstorage/storage/sql/queries_generic.py]: Since the `ttl` column as a default value, I suspect we can just select `ttl` here and let the default do the right thing, rather than trying to `COALESCE` it to a specific non-`NULL` value.

File: syncstorage/storage/sql/queries_generic.py [hunk 34]
```diff
@@ -92,6 +93,46 @@
...
+                     "FROM %(bui)s "\
+                     "WHERE batch = :batch "\
```
Comment (rfk) [inline syncstorage/storage/sql/queries_generic.py]: now that we have a composite primary key, we need to add `userid` to the `WHERE` clause here.

File: syncstorage/tests/functional/test_storage.py [hunk 35]
```diff
@@ -1286,6 +1287,97 @@ def test_pagination_with_newer_and_sort_by_oldest(self):
                 self.assertEquals(sorted(int(item['id']) for item in items),
                                   range(start + 1, NUM_ITEMS))
 
+    def assertCloseEnough(self, val1, val2, delta=0.05):
+        if abs(val1 - val2) < delta:
+            return True
+        raise AssertionError("abs(%.2f - %.2f) = %.2f > %.2f"
+                             % (val1, val2, abs(val1 - val2), delta))
```
Comment (rfk) [inline syncstorage/tests/functional/test_storage.py]: nice :-)

**Comment by rfk** [comment]
I left a couple comments but still need to come back for another pass through this.  Concurrently, we should also try to get this branch deployed somwhere that we can take it for a spin in the real world.

**Comment by rfk** [comment]
Closing this and bringing it back as a PR with branch in the mozilla-services repo
